{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "330d52bf",
   "metadata": {},
   "source": [
    "# ESFS example workflow - Parigi et al. 2022 mouse gut spatial transcriptomic\n",
    "In this workflow we apply ESFS to a mouse gut spatial transcriptomics dataset from Parigi et al. 2022. This case study show that ESFS is able to find a subset of genes that reveal distinct cell types in the adult gut. We then show that our ES combinatorial marker gene identification software find biolgically relevent gene expression profiles that non-negative matrix factorisation (NMF) fail to identify.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2713d6b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Data path\n",
    "path = \"/Users/radleya/The Francis Crick Dropbox/BriscoeJ/Radleya/New_ES_Packages/GSE169749_RAW/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26ab0232",
   "metadata": {},
   "source": [
    "### Import ESFS package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72d7d75c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ESFS\n",
    "from scipy.sparse import csc_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "110721f7",
   "metadata": {},
   "source": [
    "### Set python default discrete class colour palette to one with more colours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f20a187d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "Colours = px.colors.qualitative.Dark24\n",
    "Colours.remove('#222A2A') # Remove black form the color palette (personal preference).\n",
    "#Colours = np.concatenate((Colours,Colours))\n",
    "plt.rcParams['axes.prop_cycle'] = plt.cycler(color=Colours)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bd109f4",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ff8dbac",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load data\n",
    "adata = sc.read_10x_h5(path+\"GSM5213483_V19S23-097_A1_S1_raw_feature_bc_matrix.h5\")\n",
    "\n",
    "## Extract spatial transcriptomic grid coordinates\n",
    "Cords = pd.read_csv(path+\"GSM5213483_V19S23-097_A1_S1_tissue_positions_list.csv\",index_col=0, header=None)\n",
    "IJ_Cords = Cords[[2,3]]\n",
    "adata = adata[IJ_Cords.index]\n",
    "\n",
    "# adata = adata.copy()\n",
    "# adata.X = adata.X.toarray()\n",
    "# adata.var_names_make_unique()\n",
    "\n",
    "## Provide dummy sample labels\n",
    "adata.obs[\"Null_ID\"] = np.repeat(\"Null_ID\",adata.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95d3a680",
   "metadata": {},
   "source": [
    "### Basic Quality Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64babe2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove genes with less than 30 samples showing expression\n",
    "Keep_Genes = adata.var_names[np.where(np.sum(adata.X>0,axis=0) > 30)[0]]\n",
    "adata = adata[:,Keep_Genes]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2780253b",
   "metadata": {},
   "source": [
    "### Create and save a scaled version of the scRNA-seq counts matrix\n",
    "Create and save a scaled version of the scRNA-seq counts matrix where each genes has their observed expression values clipped if above a percentile threshold (default=97.5) and are then normalised by their maximum values so that all values are between 0-1, which is a requirment for the Entropy Sort Score (ESS) and Error Potential (EP) calcualtions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57beaa17",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Create_Scaled_Matrix(adata,log_scale=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18228599",
   "metadata": {},
   "source": [
    "### Calculate ESS and EP matricies\n",
    "Here we use the theory of Entropy Sorting to generate a pairwise gene correlation matrix (ESS matrix) and a correlation significance matrix (EP matrix). To speed up calculations parallel processing is implemented. To control the number of cores used for processing, vary the \"Use_Cores\" parameter which deaults to \"Use_Cores=-1\" which indicates the software will use n-1 cores, where n is the number of cores available on your machine. \n",
    "Please note that \"Use_Cores=-1\" is as special flag and \"Use_Cores=-2\" will not use n-2 cores. You should instead change \"Use_Cores\" to positive integer values of the number of cores you would like to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6994a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Parallel_Calc_ES_Matricies(adata, Secondary_Features_Label=\"Self\", save_matrices=np.array([\"ESSs\",\"EPs\"]), Use_Cores=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b41baccf",
   "metadata": {},
   "source": [
    "### Calculate gene importance rankings\n",
    "We now calculate the gene importance weights and hence gene importance rankings for the remaining genes in the dataset. For further details see our manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4833c0c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ESS_Threshold is one of two paramaters that users should prioratise when trying to optimise the ESFS\n",
    "# workflow. ESS_Threshold designates the upper threshold for edges between genes in the ESS matrix will be\n",
    "# retained. For this earlky human embryo data, an ESS_Threshold of 0 works well and is a good starting point\n",
    "# for most datasets. For other datasets, increasing the ESS_Threshold can be beneficial, as demonstrated in\n",
    "# our Delile et al. 2019 neural tube scRNA-seq data example workflow.\n",
    "ESS_Threshold = 0.01\n",
    "\n",
    "### An array of known important genes for early human embryo development so that we may use them as a\n",
    "# reference point for how genes are beign ranked/grouped. Replace these genes with those you are\n",
    "# interested in in your data.\n",
    "Known_Important_Genes = np.array([\"Slc51a\",\"Tagln\",\"Pcp4\"])\n",
    "\n",
    "### Run the ES_Rank_Genes function while using \"Exclude_Genes=Exclude_Batch_Effect_Genes\" to exclude possible\n",
    "# batch effect genes and \"Known_Important_Genes=Known_Important_Genes\" as reference point genes.\n",
    "adata = ESFS.ES_Rank_Genes(adata,ESS_Threshold,Known_Important_Genes=Known_Important_Genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f8aeb58",
   "metadata": {},
   "source": [
    "### Visualise clustering of top ranked genes in a UMAP generated from their pairwise ESS scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "968ef9a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Visualise top ranked genes graph\n",
    "# Num_Top_Ranked_Genes is the second of the two paramaters that users should focus on when optimising\n",
    "# the ESFS workflow for their data. Values between 3000-4000 are a typically good place to start.\n",
    "Num_Top_Ranked_Genes = 8000\n",
    "\n",
    "# The \"Clustering\" paramter can be set to an integer value if you wish to cluster genes in the UMAP with Kmeans clustering.\n",
    "# Set \"Clustering='hdbscan'\" for automated density based clustering, and \"Clustering=None\" for no clustering.\n",
    "Top_ESS_Genes, Gene_Clust_Labels, Gene_Embedding = ESFS.Plot_Top_Ranked_Genes_UMAP(adata,Num_Top_Ranked_Genes,Clustering=\"hdbscan\",Known_Important_Genes=Known_Important_Genes)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d24a194",
   "metadata": {},
   "source": [
    "### Visualise the cell clustering UMAPs for each of the groupings of genes identified in the previous step\n",
    "\n",
    "This is stage of the worfklow is a critical step for trying out different input parameters to try and optimise the final results. The 2 main paramters to vary are the \"Num_Top_Ranked_Genes\" and the \"Clustering\" inputs for the Plot_Top_Ranked_Genes_UMAP function.\n",
    "\n",
    "\"Num_Top_Ranked_Genes\" wil take the top ranked genes according to the cESFW algorithm. We recommend starting at around 3000-4000 genes and tweaking the \"Clustering\" paramater before trying higher or lower values of \"Num_Top_Ranked_Genes\".\n",
    "\n",
    "The \"Clustering\" can either be an integer number, \"hdbscan\" or \"None\". An integer number will tell the algorithm to use Kmeans clustering with that number of clusters. \"hdbscan\" will use the hdbscan densitiy based clustering algorithm to automatically identify an optimal number of cluster according the gene UMAP embedding space.\n",
    "\n",
    "In breif, you are seeking a gene cluster or combination of gene cluster that reveal biological structure of interest by excluding clusters of genes that contribute a large amount of noise to downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8b600f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_Cluster_Embeddings, Gene_Cluster_Selected_Genes = ESFS.Get_Gene_Cluster_Cell_UMAPs(adata,Gene_Clust_Labels,Top_ESS_Genes,n_neighbors=50,min_dist=0.1,log_transformed=False)\n",
    "\n",
    "ESFS.Plot_Gene_Cluster_Cell_UMAPs(adata,Gene_Cluster_Embeddings,Gene_Cluster_Selected_Genes,Cell_Label=\"Null_ID\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f11b144",
   "metadata": {},
   "source": [
    "### If you'd like to generate the cell UMAP for a specific cluster of a combination of clusters, use the specific_cluster parameter\n",
    "\n",
    "For this workflow we find cluster 2 to be most informative of neural tube progenitor to neuron differentation and choose plot the resulting UMAP from these genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf90a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_Cluster_Embeddings, Gene_Cluster_Selected_Genes = ESFS.Get_Gene_Cluster_Cell_UMAPs(adata,Gene_Clust_Labels,Top_ESS_Genes,specific_cluster=[6,3],n_neighbors=50,min_dist=0.1,log_transformed=False)\n",
    "\n",
    "ESFS.Plot_Gene_Cluster_Cell_UMAPs(adata,Gene_Cluster_Embeddings,Gene_Cluster_Selected_Genes,Cell_Label=\"Null_ID\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f514f680",
   "metadata": {},
   "source": [
    "We can also look at the gene expression of a particular gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bcd1c23",
   "metadata": {},
   "outputs": [],
   "source": [
    "ESFS.Plot_Gene_Cluster_Cell_UMAPs(adata,Gene_Cluster_Embeddings,Gene_Cluster_Selected_Genes,Cell_Label=\"Muc2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6232667f",
   "metadata": {},
   "source": [
    "### Finding marker genes\n",
    "An important task in scRNA-seq analysis is to find marker genes for distinct populations. Commonly this is acheived by using statistical tests to perform differential expressed gene (DEG) analysis. However, as discussed in our manuscript, this process is limited by the requirement to identify a set of discrete non-overlapping clusters. Here we provide code to efficiently identify genes expression profiles enriched in the combinatorial cluster space of an intentionally over clustered dataset. This is possible because ES provides a mathermatically rigorous way to turn the intractable combinatorial cluster problem into a linearly complex problem, solvable in a minutes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b63132de",
   "metadata": {},
   "source": [
    "We start by intentionally overclustering the data. To this we will use the Leiden clustering algorithm provided by the Scanpy package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a191e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sub-set the data to the genes identified by ESFS in the previous steps\n",
    "sub_adata = adata[:, Gene_Cluster_Selected_Genes[0]]\n",
    "\n",
    "# Perform Leiden clustering with a high \"resolution\" value\n",
    "sc.pp.neighbors(sub_adata, n_neighbors=50, n_pcs=0, metric = \"correlation\")\n",
    "sc.tl.umap(sub_adata,min_dist=0.1)\n",
    "sc.tl.leiden(sub_adata,resolution = 10)\n",
    "\n",
    "# Visualise the clustering\n",
    "sc.pl.umap(sub_adata, color=['leiden'],legend_fontsize=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c54354ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the cluster labels and add them to the adata object\n",
    "Leiden_Clusts = np.asarray(sub_adata.obs['leiden'])\n",
    "Unique_Leiden_Clusts = np.unique(Leiden_Clusts)\n",
    "np.min(np.unique(Leiden_Clusts,return_counts=True)[1])\n",
    "sample_labels = \"Leiden_Clusts\"\n",
    "adata.obs[sample_labels] = Leiden_Clusts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a08c5e9b",
   "metadata": {},
   "source": [
    "Now we use the Leiden clusters as a set of secondary features to identify sets of samples that maximise the correlation of feature/gene in adata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37af797",
   "metadata": {},
   "outputs": [],
   "source": [
    "Secondary_Features_Label = \"Leiden_Clusts_Secondary_Features\"\n",
    "Leiden_Clusts_Secondary_Features = csc_matrix(pd.get_dummies(Leiden_Clusts).astype(\"f\"))\n",
    "adata.obsm[Secondary_Features_Label] = Leiden_Clusts_Secondary_Features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3cd6f6c",
   "metadata": {},
   "source": [
    "We use Parallel_Calc_ES_Matricies to calculate the required metrics for ES combinatorial marker gene identification (save_matrices=np.array([\"ESSs\",\"SGs\"])). \n",
    "Secondary_Features_Label designates a prefix for saving the ESS and SG outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a638336b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Parallel_Calc_ES_Matricies(adata, Secondary_Features_Label=Secondary_Features_Label, save_matrices=np.array([\"ESSs\",\"SGs\"]), Use_Cores=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f3fce94",
   "metadata": {},
   "source": [
    "With the ESS and SG matricies saved to adata, we can use the Find_Max_ESSs function to identify which combination of Leiden Clusters maximises the correlation (ESS) or each feature/gene in adata. The most important output of Find_Max_ESSs is a 2D array where each row is the samples of adata and each column is a feature representing the identified combination of Leiden Clusters that maximises the ESS score of the respective feature/gene column in adata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d156080",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Find_Max_ESSs(adata,Secondary_Features_Label)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff55180",
   "metadata": {},
   "source": [
    "The 2D array of combined Leiden Clusters that maximise adata feature/gene ESSs is then used by Parallel_Calc_ES_Matricies as secondary features to obtain the ESS scores of each feature/gene in adata against each identified structure maximising feature. This creates a 2D pairwise adata feature/gene Vs coars grain combined Leiden Cluster matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12ec7870",
   "metadata": {},
   "outputs": [],
   "source": [
    "Secondary_Features_Label = \"Leiden_Clusts_Secondary_Features_Max_ESS_Features\"\n",
    "adata = ESFS.Parallel_Calc_ES_Matricies(adata, Secondary_Features_Label=Secondary_Features_Label, save_matrices=np.array([\"ESSs\"]), Use_Cores=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea90a46b",
   "metadata": {},
   "source": [
    "Now that we have a coarse grain represenation of where gene structure is maximised in different regions of the data, we can use Find_Minimal_Combinatorial_Gene_Set to identify a minimal set of clusters/genes that captures that most unique/non-overlapping structure in the data. This minimal gene set may be thought of as an optimised set of unsuperised marker genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d5bb23a",
   "metadata": {},
   "outputs": [],
   "source": [
    "Secondary_Features_Label\n",
    "Num_Genes = 50\n",
    "Chosen_Clusters, Chosen_Genes, Chosen_Pairwise_ESSs = ESFS.Find_Minimal_Combinatorial_Gene_Set(adata,Num_Genes,Secondary_Features_Label,Resolution=0.75,Num_Reheats=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daeb2a40",
   "metadata": {},
   "source": [
    "We can use knn_Smooth_Gene_Expression to get a knn mean smoothed representation of the data. We only use this for visualisation purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dac6dac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.knn_Smooth_Gene_Expression(adata, Gene_Cluster_Selected_Genes[0], knn=30, metric='correlation', log_scale=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c60f6a20",
   "metadata": {},
   "source": [
    "### ES Marker Genes Vs. NMF\n",
    "To demonstrate the ability of our ES marker gene software to identify distinct gene expression patterns in single cell RNA sequencing data, we compare it against non-negative matrix factorisation (NMF)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30efad80",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from scipy.optimize import linear_sum_assignment\n",
    "from sklearn.decomposition import NMF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df7c293c",
   "metadata": {},
   "source": [
    "\n",
    "First we provide the same data used to generate the above UMAP to NMF and ask it to find the same number of latent factors as we did with Find_Minimal_Combinatorial_Gene_Set (N=50)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b2e23df",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = NMF(n_components=50, init='nndsvd', max_iter=500)\n",
    "U = model.fit_transform(adata[:,Gene_Cluster_Selected_Genes[0]].layers[\"Scaled_Counts\"])\n",
    "V = model.components_"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "897fea43",
   "metadata": {},
   "source": [
    "To obtain the set of latent features from our ES marker gene algorithm we take the smoothed expression of the genes identified by Find_Minimal_Combinatorial_Gene_Set (Chosen_Genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4042a96c",
   "metadata": {},
   "outputs": [],
   "source": [
    "A = adata[:,Chosen_Genes].layers[\"Scaled_Counts\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62dc8ba3",
   "metadata": {},
   "source": [
    "No we match the most similar latent features between the ES features and NMF, while only allowing a feature to be paried with another feature once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9435619b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assume A and U are shape (4992, 30)\n",
    "A_T = A.T  # shape (30, 4992)\n",
    "U_T = U.T  # shape (30, 4992)\n",
    "# Step 1: Compute similarity matrix\n",
    "similarity_matrix = cosine_similarity(A_T, U_T)  # shape (30, 30)\n",
    "# Step 2: Convert to cost matrix (negate similarity since we want to maximize)\n",
    "cost_matrix = -similarity_matrix\n",
    "# Step 3: Solve assignment problem (Hungarian algorithm)\n",
    "row_ind, col_ind = linear_sum_assignment(cost_matrix)\n",
    "# Now: A[:, row_ind[i]] matches uniquely to U[:, col_ind[i]]\n",
    "matches = list(zip(row_ind, col_ind))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a7e754b",
   "metadata": {},
   "source": [
    "When we plot each of the pairs, we find that while in some cases both ES marker gene selection and NMF are able to identify the same cell types (Comprisons 1), there are clear cases where NMF fails to find important marker gene expression profiles that ES marker gene identification finds (Comprisons 2). Below we present some key examples. See our manuscript for further details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92892ba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import IFrame, display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca9b2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(IFrame(src=\"/Users/radleya/The Francis Crick Dropbox/BriscoeJ/Radleya/New_ES_Packages/ESFS_Git/ESFS/Example_Workflows/Gut_Spatial_Transcriptomics_Example/Comprisons 1.pdf\", width=800, height=600))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc97484a",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(IFrame(src=\"/Users/radleya/The Francis Crick Dropbox/BriscoeJ/Radleya/New_ES_Packages/ESFS_Git/ESFS/Example_Workflows/Gut_Spatial_Transcriptomics_Example/Comprisons 2.pdf\", width=800, height=600))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cf0160b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# i = 0\n",
    "# # for i in np.arange(len(matches)):\n",
    "#     #\n",
    "# Pair = matches[i]\n",
    "# #\n",
    "# plt.figure(figsize=(14,7))\n",
    "# #\n",
    "# Gene = Chosen_Genes[Pair[0]]\n",
    "# Exp = np.asarray(adata[:,Gene].layers[\"Smoothed_Expressions\"].A).T[0]\n",
    "# plt.subplot(1,2,1)\n",
    "# plt.title(Gene,fontsize=18)\n",
    "# Grid = np.zeros((np.max(IJ_Cords[2]+1),np.max(IJ_Cords[3]+1)))\n",
    "# Grid[Cords] = Exp\n",
    "# plt.imshow(Grid,cmap=\"magma\",aspect=\"auto\")  \n",
    "# plt.xticks([])\n",
    "# plt.yticks([])  \n",
    "# #\n",
    "# plt.subplot(1,2,2)\n",
    "# Factor = Pair[1]\n",
    "# Values = U[:,Factor]\n",
    "# plt.title(Factor,fontsize=18)\n",
    "# Grid = np.zeros((np.max(IJ_Cords[2]+1),np.max(IJ_Cords[3]+1)))\n",
    "# Grid[Cords] = Values\n",
    "# plt.imshow(Grid,cmap=\"magma\",aspect=\"auto\")  \n",
    "# plt.xticks([])\n",
    "# plt.yticks([])  \n",
    "# #\n",
    "# plt.subplots_adjust(0.02,0.02,0.98,0.9)\n",
    "# # plt.savefig(\"/Users/radleya/The Francis Crick Dropbox/BriscoeJ/Radleya/ESFS_Paper/ESFS Figures/Paragi2020_Plots/Matched_Plots/\" + Gene + \".png\",dpi=600)\n",
    "# # plt.close()\n",
    "\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "093576a2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
