{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ESFS example workflow - Delile et al. 2019 mouse neural tube development scRNA-sequencing data\n",
    "In this workflow we apply ESFS to a mouse embryo neural tube developmental dataset from Delile et al. 2019. As demonstrated in the UMAP below and detailed in our manuscript, ESFS allow us to identify high resultion developmental trajectories and cell states that failed to be identified by conventional scRNA-seq workflows such as Scanpy and Seurat.\n",
    "\n",
    "In this workflow ESFS is used to perform unsupervised feature selection to subset the mouse neural tube dataset from 19179 down to 879 genes. Subsetting down to the 879 genes identified by ESFS produces a UMAP displaying biologically relevent trajactories and cell types without needing apply PCA or data smoothing/imputation, indicating that ESFS has identified a set of highly biologically informative genes, which in turn facilitates improved downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Data path\n",
    "path = \"/Users/radleya/The Francis Crick Dropbox/BriscoeJ/Radleya/New_ES_Packages/Delile_Data/\"\n",
    "from IPython.display import Image, display\n",
    "display(Image(filename=path+\"Delile_2019_ESFS_UMAP.png\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import ESFS package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ESFS\n",
    "from scipy.sparse import csc_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set python default discrete class colour palette to one with more colours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "Colours = px.colors.qualitative.Dark24\n",
    "Colours.remove('#222A2A') # Remove black form the color palette (personal preference).\n",
    "#Colours = np.concatenate((Colours,Colours))\n",
    "plt.rcParams['axes.prop_cycle'] = plt.cycler(color=Colours)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import early human embryo scRNA-seq data and sample info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load data\n",
    "adata = ad.read_h5ad(path+\"ESFS_Delile_2019.h5ad\")\n",
    "# adata = ad.read_h5ad(path+\"Delile_2019.h5ad\")\n",
    "# adata = ad.read_h5ad(path+\"DelileDF.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create and save a scaled version of the scRNA-seq counts matrix\n",
    "Create and save a scaled version of the scRNA-seq counts matrix where each genes has their observed expression values clipped if above a percentile threshold (default=97.5) and are then normalised by their maximum values so that all values are between 0-1, which is a requirment for the Entropy Sort Score (ESS) and Error Potential (EP) calcualtions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Create_Scaled_Matrix(adata,log_scale=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate ESS and EP matricies\n",
    "Here we use the theory of Entropy Sorting to generate a pairwise gene correlation matrix (ESS matrix) and a correlation significance matrix (EP matrix). To speed up calculations parallel processing is implemented. To control the number of cores used for processing, vary the \"Use_Cores\" parameter which deaults to \"Use_Cores=-1\" which indicates the software will use n-1 cores, where n is the number of cores available on your machine. \n",
    "Please note that \"Use_Cores=-1\" is as special flag and \"Use_Cores=-2\" will not use n-2 cores. You should instead change \"Use_Cores\" to positive integer values of the number of cores you would like to use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Parallel_Calc_ES_Matricies(adata, Secondary_Features_Label=\"Self\", save_matrices=np.array([\"ESSs\",\"EPs\"]), Use_Cores=-1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata.write_h5ad(path+\"Delile_2019.h5ad\",compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata = ad.read_h5ad(path+\"Delile_2019.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate gene importance rankings\n",
    "We now calculate the gene importance weights and hence gene importance rankings for the remaining genes in the dataset. For further details see our manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### ESS_Threshold is one of two paramaters that users should prioratise when trying to optimise the ESFS\n",
    "# workflow. ESS_Threshold designates the upper threshold for edges between genes in the ESS matrix will be\n",
    "# retained. For this earlky human embryo data, an ESS_Threshold of 0 works well and is a good starting point\n",
    "# for most datasets. For other datasets, increasing the ESS_Threshold can be beneficial, as demonstrated in\n",
    "# our Delile et al. 2019 neural tube scRNA-seq data example workflow.\n",
    "ESS_Threshold = 0.01\n",
    "\n",
    "### An array of known important genes for early human embryo development so that we may use them as a\n",
    "# reference point for how genes are beign ranked/grouped. Replace these genes with those you are\n",
    "# interested in in your data.\n",
    "Known_Important_Genes = np.load(path+\"Delile2019_Knowledge_Matrix_Genes.npy\")\n",
    "### Run the ES_Rank_Genes function while using \"Exclude_Genes=Exclude_Batch_Effect_Genes\" to exclude possible\n",
    "# batch effect genes and \"Known_Important_Genes=Known_Important_Genes\" as reference point genes.\n",
    "adata = ESFS.ES_Rank_Genes(adata,ESS_Threshold,Known_Important_Genes=Known_Important_Genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualise clustering of top ranked genes in a UMAP generated from their pairwise ESS scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Visualise top ranked genes graph\n",
    "# Num_Top_Ranked_Genes is the second of the two paramaters that users should focus on when optimising\n",
    "# the ESFS workflow for their data. Values between 3000-4000 are a typically good place to start.\n",
    "Num_Top_Ranked_Genes = 3000\n",
    "\n",
    "# The \"Clustering\" paramter can be set to an integer value if you wish to cluster genes in the UMAP with Kmeans clustering.\n",
    "# Set \"Clustering='hdbscan'\" for automated density based clustering, and \"Clustering=None\" for no clustering.\n",
    "Top_ESS_Genes, Gene_Clust_Labels, Gene_Embedding = ESFS.Plot_Top_Ranked_Genes_UMAP(adata,Num_Top_Ranked_Genes,Clustering=\"hdbscan\",Known_Important_Genes=Known_Important_Genes)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualise the cell clustering UMAPs for each of the groupings of genes identified in the previous step\n",
    "\n",
    "This is stage of the worfklow is a critical step for trying out different input parameters to try and optimise the final results. The 2 main paramters to vary are the \"Num_Top_Ranked_Genes\" and the \"Clustering\" inputs for the Plot_Top_Ranked_Genes_UMAP function.\n",
    "\n",
    "\"Num_Top_Ranked_Genes\" wil take the top ranked genes according to the cESFW algorithm. We recommend starting at around 3000-4000 genes and tweaking the \"Clustering\" paramater before trying higher or lower values of \"Num_Top_Ranked_Genes\".\n",
    "\n",
    "The \"Clustering\" can either be an integer number, \"hdbscan\" or \"None\". An integer number will tell the algorithm to use Kmeans clustering with that number of clusters. \"hdbscan\" will use the hdbscan densitiy based clustering algorithm to automatically identify an optimal number of cluster according the gene UMAP embedding space.\n",
    "\n",
    "In breif, you are seeking a gene cluster or combination of gene cluster that reveal biological structure of interest by excluding clusters of genes that contribute a large amount of noise to downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_Cluster_Embeddings, Gene_Cluster_Selected_Genes = ESFS.Get_Gene_Cluster_Cell_UMAPs(adata,Gene_Clust_Labels,Top_ESS_Genes,n_neighbors=50,min_dist=0.1,log_transformed=True)\n",
    "\n",
    "ESFS.Plot_Gene_Cluster_Cell_UMAPs(adata,Gene_Cluster_Embeddings,Gene_Cluster_Selected_Genes,Cell_Label=\"domain\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If you'd like to generate the cell UMAP for a specific cluster of a combination of clusters, use the specific_cluster parameter\n",
    "\n",
    "For this workflow we find cluster 2 to be most informative of neural tube progenitor to neuron differentation and choose plot the resulting UMAP from these genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_Cluster_Embeddings, Gene_Cluster_Selected_Genes = ESFS.Get_Gene_Cluster_Cell_UMAPs(adata,Gene_Clust_Labels,Top_ESS_Genes,specific_cluster=[2],n_neighbors=40,min_dist=0.1,log_transformed=True)\n",
    "\n",
    "ESFS.Plot_Gene_Cluster_Cell_UMAPs(adata,Gene_Cluster_Embeddings,Gene_Cluster_Selected_Genes,Cell_Label=\"domain\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As outlined in our mansucript, using this high resolution UMAP embedding from ESFS, we were able to perform unsupervised Leiden clustering and re-annotate mislabled cells from previous analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ESFS.Plot_Gene_Cluster_Cell_UMAPs(adata,Gene_Cluster_Embeddings,Gene_Cluster_Selected_Genes,Cell_Label=\"This_Paper_Reannoations\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding marker genes\n",
    "An important task in scRNA-seq analysis is to find marker genes for distinct populations. Commonly this is acheived by using statistical tests to perform differential expressed gene (DEG) analysis. However, as discussed in our manuscript, this process is limited by the requirement to identify a set of discrete non-overlapping clusters. Here we provide code to efficiently identify genes expression profiles enriched in the combinatorial cluster space of an intentionally over clustered dataset. This is possible because ES provides a mathermatically rigorous way to turn the intractable combinatorial cluster problem into a linearly complex problem, solvable in a minutes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start by taking the intentionally overclustered sample labels, Leiden_Clusts, and converting them into a one-hot representation. These one-hot encoded clusters will serve as \"Secondary_Features\" to calculate the Entropy Sort Score (ESS) and Sort Gain (SG) metrics when compared against every feature/gene in adata.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.sparse import csc_matrix\n",
    "\n",
    "Leiden_Clusts = np.load(\"/Users/radleya/Desktop/Leiden_Clusts.npy\",allow_pickle=True)\n",
    "adata.obs[\"Leiden_Clusts\"] = Leiden_Clusts\n",
    "Fixed_Features = csc_matrix(pd.get_dummies(Leiden_Clusts).astype(\"f\"))\n",
    "\n",
    "Secondary_Features_Label = \"Leiden_Clusts_Secondary_Features\"\n",
    "Leiden_Clusts_Secondary_Features = csc_matrix(pd.get_dummies(Leiden_Clusts).astype(\"f\"))\n",
    "adata.obsm[Secondary_Features_Label] = Leiden_Clusts_Secondary_Features\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we use Parallel_Calc_ES_Matricies to calculate the required metrics for ES combinatorial marker gene identification (save_matrices=np.array([\"ESSs\",\"SGs\"])). \n",
    "Secondary_Features_Label designates a prefix for saving the ESS and SG outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Parallel_Calc_ES_Matricies(adata, Secondary_Features_Label=Secondary_Features_Label, save_matrices=np.array([\"ESSs\",\"SGs\"]), Use_Cores=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the ESS and SG matricies saved to adata, we can use the Find_Max_ESSs function to identify which combination of Leiden Clusters maximises the correlation (ESS) or each feature/gene in adata. The most important output of Find_Max_ESSs is a 2D array where each row is the samples of adata and each column is a feature representing the identified combination of Leiden Clusters that maximises the ESS score of the respective feature/gene column in adata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.Find_Max_ESSs(adata,Secondary_Features_Label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The 2D array of combined Leiden Clusters that maximise adata feature/gene ESSs is then used by Parallel_Calc_ES_Matricies as secondary features to obtain the ESS scores of each feature/gene in adata against each identified structure maximising feature. This creates a 2D pairwise adata feature/gene Vs coars grain combined Leiden Cluster matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Secondary_Features_Label = \"Leiden_Clusts_Secondary_Features_Max_ESS_Features\"\n",
    "adata = ESFS.Parallel_Calc_ES_Matricies(adata, Secondary_Features_Label=Secondary_Features_Label, save_matrices=np.array([\"ESSs\"]), Use_Cores=-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have a coarse grain represenation of where gene structure is maximised in different regions of the data, we can use Find_Minimal_Combinatorial_Gene_Set to identify a minimal set of clusters/genes that captures that most unique/non-overlapping structure in the data. This minimal gene set may be thought of as an optimised set of unsuperised marker genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Secondary_Features_Label\n",
    "Num_Genes = 300\n",
    "Chosen_Clusters, Chosen_Genes, Chosen_Pairwise_ESSs = ESFS.Find_Minimal_Combinatorial_Gene_Set(adata,Num_Genes,Secondary_Features_Label,Resolution=0.75,Num_Reheats=3)\n",
    "\n",
    "# import seaborn as sns\n",
    "# sns.clustermap(Chosen_Pairwise_ESSs)\n",
    "# plt.savefig(\"/Users/radleya/The Francis Crick Dropbox/BriscoeJ/Radleya/ESFS_Paper/ESFS Figures/Paragi2020_Plots/Matched_Plots/\" + \"50_Factor_Heatmap\" + \".png\",dpi=600)\n",
    "# plt.close()\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use knn_Smooth_Gene_Expression to get a knn mean smoothed representation of the data. We only use this for visualisation purposes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ESFS.knn_Smooth_Gene_Expression(adata, Gene_Cluster_Selected_Genes[0], knn=30, metric='correlation', log_scale=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ESFS.Plot_Gene_Cluster_Cell_UMAPs(adata,Gene_Cluster_Embeddings,Gene_Cluster_Selected_Genes,Cell_Label=\"Sox2\") # Hypoblast marker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.animation as animation\n",
    "from IPython.display import Image, display\n",
    "import tempfile\n",
    "import os\n",
    "\n",
    "def Display_Chosen_Genes_gif(Embedding,Chosen_Genes,adata):\n",
    "    # Initialize figure and axis\n",
    "    fig, ax = plt.subplots(figsize=(6, 6))\n",
    "\n",
    "    # Function to update each frame\n",
    "    def update(i):\n",
    "        ax.clear()  # Clear previous frame\n",
    "        Gene = Chosen_Genes[i]\n",
    "        Exp = adata[:,Gene].layers[\"Smoothed_Expression\"].A.ravel()\n",
    "        Order = np.argsort(Exp)\n",
    "        #\n",
    "        ax.set_title(Gene, fontsize=22)\n",
    "        Vmax = np.percentile(Exp, 99)\n",
    "        if Vmax == 0:\n",
    "            Vmax = np.max(Exp)\n",
    "        scatter = ax.scatter(Embedding[Order, 0], Embedding[Order, 1], c=Exp[Order], s=2, vmax=Vmax, cmap=\"seismic\")\n",
    "        #\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "        fig.subplots_adjust(0.02, 0.02, 0.98, 0.9)\n",
    "\n",
    "        return scatter,\n",
    "    #\n",
    "    # Create animation\n",
    "    Num_Frames = Chosen_Genes.shape[0]\n",
    "    ani = animation.FuncAnimation(fig, update, frames=Num_Frames, interval=500)\n",
    "    #\n",
    "    plt.close(\"all\")\n",
    "    #\n",
    "    # Save animation as GIF\n",
    "    with tempfile.TemporaryDirectory() as tmpdir:\n",
    "        gif_path = os.path.join(tmpdir, \"gene_expression.gif\")\n",
    "        ani.save(gif_path, writer=animation.PillowWriter(fps=0.5))\n",
    "        # Display the GIF inline in Jupyter\n",
    "        with open(gif_path, \"rb\") as f:\n",
    "            display(Image(data=f.read(), format='png'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Embedding = Gene_Cluster_Embeddings[0]\n",
    "Display_Chosen_Genes_gif(Embedding,Chosen_Genes,adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combinatorial_Label_Info = adata.varm[\"Leiden_Clusts_Secondary_Features_Max_Combinatorial_ESSs\"]\n",
    "# Plot_Genes = Chosen_Genes[np.argsort(Combinatorial_Label_Info.loc[Chosen_Genes][\"Max_ESSs\"])].copy()\n",
    "Cluster_ESSs = adata.varm['Leiden_Clusts_Secondary_Features_Max_ESS_Features_ESSs'][Chosen_Clusters,:]\n",
    "col_max = np.max(Cluster_ESSs, axis=0, keepdims=True)\n",
    "\n",
    "# Use broadcasting to create a mask where the max values are located\n",
    "mask = Cluster_ESSs == col_max\n",
    "\n",
    "# Apply the mask to keep only max values; all others become 0\n",
    "Cluster_ESSs = Cluster_ESSs * mask"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
